### ISSUE TITLE
0.12 beta2 init fails when use of multiple symlinks

### ISSUE THREAD

**Issue by dblooman** [hashicorp/terraform#21060]
### Terraform Version

```
Terraform v0.12.0-beta2
```

### Terraform Configuration Files
https://github.com/dblooman/terraform-bug

In this repo, `cd staging/services/demo_service` and `terraform init`

### Debug Output
```bash
2019/04/19 15:28:23 [INFO] Terraform version: 0.12.0 beta2 
2019/04/19 15:28:23 [INFO] Go runtime version: go1.11.5
2019/04/19 15:28:23 [INFO] CLI args: []string{"/usr/local/bin/terraform2", "init"}
2019/04/19 15:28:23 [DEBUG] Attempting to open CLI config file: /Users/dblooman/.terraformrc
2019/04/19 15:28:23 [DEBUG] File doesn't exist, but doesn't need to. Ignoring.
2019/04/19 15:28:23 [INFO] CLI command args: []string{"init"}
2019/04/19 15:28:23 [TRACE] ModuleInstaller: installing child modules for . into .terraform/modules
Initializing modules...
2019/04/19 15:28:23 [DEBUG] Module installer: begin testing
2019/04/19 15:28:23 [TRACE] ModuleInstaller: Module installer: testing <nil> already installed in ../modules/demo
2019/04/19 15:28:23 [DEBUG] Module installer: begin testing.ami
2019/04/19 15:28:23 [TRACE] ModuleInstaller: testing.ami has local path "../../aws_example"
2019/04/19 15:28:23 [TRACE] ModuleInstaller: testing.ami uses directory from parent: ../aws_example
2019/04/19 15:28:23 [DEBUG] Module installer: testing.ami installed at ../aws_example
2019/04/19 15:28:23 [TRACE] modsdir: writing modules manifest to .terraform/modules/modules.json
- testing.ami in ../aws_example

Error: Failed to read module directory

Module directory ../aws_example does not exist or cannot be read.


Error: Failed to read module directory

Module directory ../aws_example does not exist or cannot be read.

```

### Expected Behavior
I should be able to terraform init and get modules as with current latest stable terraform

### Actual Behavior
Terraform doesn't init, fails to find module

### Steps to Reproduce
clonse https://github.com/dblooman/terraform-bug

In this repo, `cd staging/services/demo_service` and `terraform init`

**Comment by mildwonkey** [comment]
Hi @dblooman !
Thank you for reporting this, and providing a reproduction case! This will help me nail down where exactly terraform is looking for the relative paths to the modules.

I'm going to add a few notes here from my triaging - these are mostly meant to remind me what to look at next, so please don't worry that these notes don't provide an answer yet!

Using the repository you provided, I made the following change to `modules/services/demo/main.tf` and `terraform init` worked:

```
[kristin@fairy /tmp/terraform-bug:master *]
$ git diff modules/services/demo/main.tf
diff --git a/modules/services/demo/main.tf b/modules/services/demo/main.tf
index c8419ad..0d003c4 100644
--- a/modules/services/demo/main.tf
+++ b/modules/services/demo/main.tf
@@ -4,5 +4,5 @@ resource "aws_s3_bucket" "b" {
 }

 module "ami" {
-  source = "../../aws_example"
+  source = "../../../../modules/aws_example/"
 }
```

The trace log show where terraform is looking - note that the actual path relative is  `../../../modules/aws_example`:
```
2019/04/19 13:03:28 [TRACE] ModuleInstaller: testing.ami has local path "../../../../modules/aws_example/"
2019/04/19 13:03:28 [TRACE] ModuleInstaller: testing.ami uses directory from parent: ../../../modules/aws_example
2019/04/19 13:03:28 [DEBUG] Module installer: testing.ami installed at ../../../modules/aws_example
```

**Comment by mildwonkey** [comment]
There is a workaround that may or may not work for your actual situation, but it does work in the example repository. I changed the symlink from `modules/services` to the top-level `modules` directory:

Before:
```
$ ls -la modules
lrwxr-xr-x  1 kristin  wheel  22 Apr 19 13:31 modules -> ../../modules/services
```

After:
```
$ ls -la modules
lrwxr-xr-x  1 kristin  wheel  14 Apr 19 13:29 modules -> ../../modules/
```

And updated `main.tf` to point to the module in `./modules/services/demo`:

```hcl
 provider "aws" {}

 module "testing" {
    source = "../modules/services/demo"
 }
```

`terraform init` ran successfully after that.

**Comment by dblooman** [comment]
Yeah, I guess that solution would work, I'm not sure if we are the edge case and should change or if there is a solution that means we don't have to, I will do some more testing to see how our tooling and scripts react to this change

**Comment by mildwonkey** [comment]
@dblooman no worries - this is a bug and I fully expect to have this fixed before 0.12 is released.

**Comment by dblooman** [comment]
Thanks 👍
